File: 0516-longest-palindromic-subsequence/0516-longest-palindromic-subsequence.py
class Solution:
    def longestPalindromeSubseq(self, s: str) -> int:
        s1 = s[::-1]
        print(s)
        print(s1)
        n = len(s)
        dp =[ [0 for i in range (n+1)] for j in range (n+1)]
        for i in range (1,n+1):
            for j in range (1,n+1):
                if(s[i-1]==s1[j-1]):
                    dp[i][j]= 1+dp[i-1][j-1]
                else :
                    dp[i][j] = max(dp[i-1][j],dp[i][j-1])
        return dp[n][n]

File: 0516-longest-palindromic-subsequence/0516-longest-palindromic-subsequence.cpp
class Solution {
public:
    map<pair<int,int>,int>mp;
    int com( string &s1,string &s2, int m,int n){
        if(mp.find(make_pair(m,n))!=mp.end()){
            return mp[make_pair(m,n)];
        }
        if(m==0 or n==0){
            return mp[make_pair(m,n)]=0;
        }
        if(s1[m-1]==s2[n-1]){
            return mp[make_pair(m,n)]=1 + com(s1,s2,m-1,n-1);
        }
        return mp[make_pair(m,n)]=max(com(s1,s2,m-1,n),com(s1,s2,m,n-1));
    }
    int longestPalindromeSubseq(string s) {
        string s1;
        for(int i =0;i<s.size();++i){
            s1+=s[i];
        }
        reverse( begin(s1),end(s1));
        return com(s,s1,s.size(),s.size());
    }
};